File: 2018/aoc2018/day17/day17.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strings"
)

type Range struct {
	x0 int
	x1 int
	y0 int
	y1 int
}

type grid [][]byte

func New(ranges []Range, minX int, maxX int, minY int, maxY int) grid {
	sizeX := maxX + 1
	var grid [][]byte
	for y := 0; y <= maxY+3; y++ {
		var row []byte
		for x := 0; x <= sizeX; x++ {
			row = append(row, '.')
		}
		grid = append(grid, row)
	}

	for _, elem := range ranges {
		for y := elem.y0; y <= elem.y1; y++ {
			for x := elem.x0; x <= elem.x1; x++ {
				grid[y][x] = '#'
			}
		}
	}
	grid[0][500] = '+'
	return grid
}

func (g grid) String() string {
	var s strings.Builder
	minX := math.MaxInt
	for _, row := range g {
		for ix, v := range row {
			if v != '.' && ix < minX {
				minX = ix
			}
		}
	}
	for _, row := range g {
		for ix, x := range row {
			if ix >= minX-1 {
				s.WriteByte(x)
			}
		}
		s.WriteByte('\n')
	}
	return s.String()
}

type Point struct {
	x int
	y int
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	minY := math.MaxInt
	minX := math.MaxInt
	maxY := math.MinInt
	maxX := math.MinInt
	var ranges []Range

	for scanner.Scan() {
		txt := scanner.Text()
		var x, y, x1, x0, y0, y1 int
		_, err := fmt.Sscanf(txt, "x=%d, y=%d..%d", &x, &y0, &y1)
		if err == nil {
			if y0 < minY {
				minY = y0
			}
			if y1 > maxY {
				maxY = y1
			}
			if x > maxX {
				maxX = x
			}
			if x < minX {
				minX = x
			}
			ranges = append(ranges, Range{x0: x, x1: x, y0: y0, y1: y1})
			continue
		}
		_, err = fmt.Sscanf(txt, "y=%d, x=%d..%d", &y, &x0, &x1)
		if err != nil {
			panic(err)
		}
		if y < minY {
			minY = y
		}
		if y > maxY {
			maxY = y
		}
		if x1 > maxX {
			maxX = x1
		}
		if x0 < minX {
			minX = x0
		}

		ranges = append(ranges, Range{x0: x0, x1: x1, y0: y, y1: y})

	}

	grid := New(ranges, minX, maxX, minY, maxY)

	escapes := func(p Point) bool {
		stack := []Point{p}
		seen := make(map[Point]bool)
		goal := p.y + 1
		for len(stack) > 0 {
			current := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			seen[current] = true
			x := current.x
			y := current.y

			if y >= goal {
				return true
			}
			down := Point{x: x, y: y + 1}
			if down.y > maxY {
				continue
			}
			if grid[down.y][down.x] == '.' || grid[down.y][down.x] == '|' {
				stack = append(stack, down)
				continue
			}
			right := Point{x: x + 1, y: y}
			left := Point{x: x - 1, y: y}
			if grid[left.y][left.x] == '.' || grid[left.y][left.x] == '|' {
				if !seen[left] {
					stack = append(stack, left)
				}
			}
			if grid[right.y][right.x] == '.' || grid[right.y][right.x] == '|' {
				if !seen[right] {
					stack = append(stack, right)
				}
			}

		}
		return false
	}
	changed := true
	for changed {
		changed = false
		current := Point{x: 500, y: 1}
		queue := []Point{current}
		seen := make(map[Point]bool)
		for len(queue) > 0 {
			current := queue[0]
			queue = queue[1:]
			y := current.y
			x := current.x
			seen[current] = true
			currentVal := grid[y][x]
			down := Point{x: x, y: y + 1}
			if down.y > maxY {
				if currentVal != '|' {
					changed = true
				}
				grid[y][x] = '|'
				continue
			}

			if grid[down.y][down.x] == '.' || grid[down.y][down.x] == '|' {
				queue = append(queue, down)
			} else {
				for _, p := range []Point{{x: x + 1, y: y}, {x: x - 1, y: y}} {
					if grid[p.y][p.x] == '.' || grid[p.y][p.x] == '|' {
						if !seen[p] {
							queue = append(queue, p)
						}
					}
				}
			}

			if escapes(current) {
				if currentVal != '|' {
					changed = true
				}
				grid[y][x] = '|'
			} else {
				if currentVal != '~' {
					changed = true
				}
				grid[y][x] = '~'
			}

		}
	}

	total := 0
	p2 := 0
	for y := minY; y <= maxY; y++ {
		for x := 0; x < len(grid[0]); x++ {
			if grid[y][x] == '|' || grid[y][x] == '~' {
				total += 1
			}
			if grid[y][x] == '~' {
				p2 += 1
			}
		}
	}
	fmt.Println(total, p2)
}

File: 2018/day17/day17.py
import fileinput
from collections import deque
import math
import re

grid = {}
lines = []
for line in fileinput.input():
    line = line.strip()
    lines.append(line)


for line in lines:
    if line.startswith("x="):
        (x,) = re.match(r"x=(\d+),", line).groups()
        y1, y2 = line.split("y=")[1].split("..")
        y1 = int(y1)
        y2 = int(y2)
        x = int(x)
        for y in range(y1, y2 + 1):
            grid[(x, y)] = "#"
    elif line.startswith("y"):
        (y,) = re.match(r"y=(\d+),", line).groups()
        y = int(y)
        x1, x2 = line.split("x=")[1].split("..")
        x1 = int(x1)
        x2 = int(x2)
        for x in range(x1, x2 + 1):
            grid[(x, y)] = "#"

minx = math.inf
maxx = -math.inf
miny = math.inf
maxy = -math.inf
for x, y in grid.keys():
    if x < minx:
        minx = x
    if y < miny:
        miny = y
    if x > maxx:
        maxx = x
    if y > maxy:
        maxy = y


def print_grid(g):
    for y in range(0, int(maxy) + 2):
        for x in range(int(minx) - 1, int(maxx) + 2):
            if y == 0 and x == 500:
                print("+", end="")

            elif (x, y) not in g:
                print(".", end="")
            else:
                print(g[(x, y)], end="")
        print()


def escapes(p: tuple[int, int]) -> bool:
    stack = [p]
    seen = set()
    goaly = p[1] + 1
    while stack:
        current = stack.pop()
        seen.add(current)
        x, y = current
        if y >= goaly:
            return True
        # try down
        down = (x, y + 1)
        if down[1] > maxy:
            continue

        if grid.get(down, ".") == "." or grid.get(down, ".") == "|":
            stack.append(down)
            continue
        right = (x + 1, y)
        if grid.get(right, ".") in (".", "|"):
            if right not in seen:
                stack.append(right)
        left = (x - 1, y)
        if grid.get(left, ".") in (".", "|"):
            if left not in seen:
                stack.append(left)
    return False


m = 0
changed = True

while changed:
    m += 1
    current = (500, 1)
    queue = deque([current])
    seen = set()
    changed = False
    while queue:
        current = queue.popleft()
        x, y = current
        seen.add(current)
        current_val = grid.get(current)

        # try down
        down = (x, y + 1)
        if down[1] > maxy:
            if current_val != "|":
                changed = True
            grid[current] = "|"
            continue

        if grid.get(down, ".") == "." or grid.get(down, ".") == "|":
            queue.append(down)
        else:
            right = (x + 1, y)
            if grid.get(right, ".") in (".", "|"):
                if right not in seen:
                    queue.append(right)
            left = (x - 1, y)
            if grid.get(left, ".") in (".", "|"):
                if left not in seen:
                    queue.append(left)

        if not escapes(current):
            if current_val != "~":
                changed = True
            grid[current] = "~"
        else:
            if current_val != "|":
                changed = True
            grid[current] = "|"

    if not changed:
        break

total = 0
retained = 0
for k, v in grid.items():
    x, y = k
    if y >= miny and y <= maxy:
        if v == "|" or v == "~":
            total += 1
    if v == "~":
        retained += 1
print(total, retained)
